### PR TITLE
Update model execution API

### PR DESCRIPTION
**Pull request by wchao1115**
1. Support optional outputs and optional input shapes. (#87)
2. Model builder interface for state separation between context state and graph state. (#87)
3. Fold the execution interface into the compilation interface for simplicity. (#87)
4. Remove quantization-related params from the operand descriptor. (#44)
5. Simplify operand type enum. (#87)
6. Update URL references of first-wave operators. (#66)
7. Update spec code samples and comments accordingly. 

@gramalingam

***
<a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/pull/94.html" title="Last updated on Oct 7, 2020, 5:51 AM UTC (24eba44)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/94/d740b23...24eba44.html" title="Last updated on Oct 7, 2020, 5:51 AM UTC (24eba44)">Diff</a>

### CODE REVIEW

File: index.bs
Comment (huningxin) [inline index.bs:206]: Do we need to support `int8` and `uint8`?

File: op_compatibility/first_wave_models.md
Comment (huningxin) [inline op_compatibility/first_wave_models.md:30]: Should the link to `#api-modelbuilder-unary` after this PR landed?

File: op_compatibility/first_wave_models.md
Comment (huningxin) [inline op_compatibility/first_wave_models.md:34]: ditto.

File: index.bs
Comment (huningxin) [inline index.bs:239]: `NamedOperands`? `Opearnds` sounds like a list.

File: index.bs
Comment (huningxin) [inline index.bs:213]: As we are going to remove the `tensor-type` values of `OperandType`, we may need to define how to define a scalar. The options are no `dimensions` member, e.g. `{type: 'float32'}` or empty `dimensions`, e.g. {type: 'float32', dimensions: []}.

**Review by huningxin — COMMENTED** [review]
@wchao1115 , thanks for your great work. I left some comments, please take a look.

File: index.bs
Comment (wchao1115) [reply index.bs:1136]: Thanks @huningxin. I can make that change.

Can we actually use `record<K,V>`? I just noticed that it's only defined in the Editor's Draft and not on [W3C WebIDL Level 1](https://www.w3.org/TR/WebIDL-1/).

File: op_compatibility/first_wave_models.md
Comment (wchao1115) [reply op_compatibility/first_wave_models.md:30]: Good catch! I updated all occurrences of #api-neuralnetworkcontext to reflect the new links.

File: index.bs
Comment (wchao1115) [reply index.bs:239]: `NamedInputs` and `NamedOutputs` for consistency?

File: index.bs
Comment (wchao1115) [reply index.bs:213]: I plan to open a new issue on scalar tensor.

File: index.bs
Comment (wchao1115) [reply index.bs]: Agreed with this suggestion. Fixed.

File: index.bs
Comment (wchao1115) [reply index.bs]: Fixed.

File: index.bs
Comment (wchao1115) [reply index.bs]: Changed.

File: index.bs
Comment (wchao1115) [reply index.bs]: Agreed. Added.

File: op_compatibility/first_wave_models.md
Comment (wchao1115) [reply op_compatibility/first_wave_models.md]: Fixed.

File: op_compatibility/first_wave_models.md
Comment (wchao1115) [reply op_compatibility/first_wave_models.md]: Fixed.

File: index.bs
Comment (huningxin) [reply index.bs]: sounds good.

File: index.bs
Comment (huningxin) [reply index.bs:215]: +1

File: index.bs
Comment (huningxin) [reply index.bs]: Good question. @anssiko , could you please chime in?

I found [WebAudio spec](https://www.w3.org/TR/webaudio/#dictdef-audioworkletnodeoptions) uses `record<K,V>` for [`AudioWorkletNodeOptions.parameterData`](https://www.w3.org/TR/webaudio/#dom-audioworkletnodeoptions-parameterdata).

**Comment by huningxin** [comment]
Another thing is whether we need to change the informative sample code where `nn` is used. e.g. for [relu](https://webmachinelearning.github.io/webnn/#api-neuralnetworkcontext-relu), the sample code is
```js
return nn.max(nn.constant(0), x);
```

File: index.bs
Comment (anssiko) [reply index.bs]: Please use https://heycam.github.io/webidl/ for WebIDL.

**Review by anssiko — APPROVED** [review]

**Review by huningxin — APPROVED** [review]
Looks good to me. Thanks @wchao1115 .

File: index.bs
Comment (huningxin) [inline index.bs:299]: should it be `builder.reshape`? it was an old issue.

File: index.bs
Comment (pyu10055) [inline index.bs:1034]: With this API, the computation is tied with the compilation. should the compile method have the inputs/outputs pair and execution can only execution the graph compiled with the input/output pair? Otherwise, the compilation might not support the execution of the sub-graph scenario?

Comment (wchao1115) [reply index.bs]: @pyu10055 Good idea. Will this work now?

File: index.bs
Comment (wchao1115) [reply index.bs]: Fixed.

**Comment by anssiko** [comment]
@pyu10055 PTAL. This is the last call for comments prior to merging this PR. Please provide your feedback by Pacific Time afternoon today.

We'd like to get this PR landed to unblock other pending changes. Thanks!

File: index.bs
Comment (huningxin) [reply index.bs]: Do we need this change?

@pyu10055 , according to https://github.com/webmachinelearning/webnn/issues/87#issuecomment-694330940, you seemed to agree the ` Promise<NamedOutputs> compute(NamedInputs inputs, optional NamedOutputs outputs);` (the ability to specify the outputs when executing) would support the sub-graph execution use case, correct?

Comment (wchao1115) [reply index.bs]: @huningxin I actually do not recommend partial execution of a compiled graph as any operator in between the input and output operands may be folded with the surrounding operators from fusions or even eliminated by constant folding. The only guaranteed executable operands are the input and the output operands of the compiled graph. Partial compilation of a model is much more robust. Once a compilation is produced for the partial model, it stays valid and can be executed as such.

Comment (huningxin) [reply index.bs]: @wchao1115 , I only mean the `compute` method now allows developers to select the output operands of a compiled graph. I didn't recommend the partial execution for any intermediate operands between inputs and outputs.

My point is that the graph building is in the context of new `ModelBuilder` object. Developers are free to create a sub-graph by `createModel` API and compile/execute it. I think it would support the sub-graph execution use case without adding inputs/outputs pair to `compile` method.

Comment (wchao1115) [reply index.bs:1034]: @huningxin and I discussed this specific change offline. We want to track partial graph execution as a separate issue. @pyu10055 you may create a new issue just for this topic if you wish. I will now complete this PR with the original set of changes as outlined in the PR description above.

**Comment by zolkis** [comment]
Good changes, helps clarity.

File: index.bs
Comment (pyu10055) [inline index.bs:295]: ideally it would be great to allow chained api for the Operands.
builder.constant(...).add(builder.constant(...))

File: index.bs
Comment (huningxin) [reply index.bs:1034]: Created https://github.com/webmachinelearning/webnn/issues/105 as follow-up.

File: index.bs
Comment (huningxin) [reply index.bs:295]: I think this is a good idea. Created https://github.com/webmachinelearning/webnn/issues/106 to follow up.

**Pull request merged**
